File: backend/supabase/functions/_shared/marketService.ts
import { sql } from "./db.ts";

const YAHOO_HEADERS = {
  "User-Agent": "Mozilla/5.0 (compatible; TemptationDashboard/1.0)",
  "Accept": "application/json",
};

const QUOTE_CACHE_TTL_SECONDS = 4;

async function yahooChart(symbol: string): Promise<any | null> {
  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${symbol}`;
  try {
    const res = await fetch(url, { headers: YAHOO_HEADERS, signal: AbortSignal.timeout(8000) });
    if (!res.ok) return null;
    return await res.json();
  } catch {
    return null;
  }
}

export async function symbolExistsOnNse(name: string): Promise<boolean> {
  const clean = name.trim().toUpperCase();
  if (!clean || clean.length > 50) return false;
  const data = await yahooChart(`${clean}.NS`);
  if (!data) return false;
  const result = data?.chart?.result;
  if (!result || !result.length) return false;
  const meta = result[0]?.meta ?? {};
  const price = meta.regularMarketPrice ?? meta.previousClose;
  return price !== null && price !== undefined;
}

export async function fetchNifty50Quote(): Promise<{ value: number | null; changePercent: number | null }> {
  const data = await yahooChart("^NSEI");
  if (!data) return { value: null, changePercent: null };
  const result = data?.chart?.result;
  if (!result || !result.length) return { value: null, changePercent: null };
  const meta = result[0]?.meta ?? {};
  const price = meta.regularMarketPrice;
  const prev = meta.previousClose ?? meta.chartPreviousClose;
  if (price === null || price === undefined || prev === null || prev === undefined || prev === 0) {
    return { value: price ?? null, changePercent: null };
  }
  const changePercent = ((price - prev) / prev) * 100;
  return { value: price, changePercent };
}

export async function fetchStockQuoteNse(name: string): Promise<number | null> {
  const clean = name.trim().toUpperCase();
  if (!clean) return null;
  const data = await yahooChart(`${clean}.NS`);
  if (!data) return null;
  const result = data?.chart?.result;
  if (!result || !result.length) return null;
  const meta = result[0]?.meta ?? {};
  const price = meta.regularMarketPrice ?? meta.previousClose;
  return price !== null && price !== undefined ? Number(price) : null;
}

/**
 * Short-TTL cache in front of fetchStockQuoteNse, backed by Postgres so the price used to
 * display (dashboard/execute reads) and the price used to auto-execute trades stay the same
 * value across separate Edge Function invocations within the TTL window (replaces the
 * Python in-process dict cache, which a stateless per-invocation model can't rely on).
 */
export async function getQuoteCached(name: string): Promise<number | null> {
  const clean = name.trim().toUpperCase();
  const rows = await sql<{ price: string | null; fetched_at: Date }[]>`
    select price, fetched_at from quote_cache where symbol = ${clean}
  `;
  if (rows.length > 0) {
    const ageSeconds = (Date.now() - rows[0].fetched_at.getTime()) / 1000;
    if (ageSeconds < QUOTE_CACHE_TTL_SECONDS) {
      return rows[0].price === null ? null : Number(rows[0].price);
    }
  }
  const price = await fetchStockQuoteNse(clean);
  await sql`
    insert into quote_cache (symbol, price, fetched_at)
    values (${clean}, ${price}, now())
    on conflict (symbol) do update set price = excluded.price, fetched_at = excluded.fetched_at
  `;
  return price;
}

File: backend/app/services/market_service.py
import time
from threading import Lock

import httpx

YAHOO_HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; TemptationDashboard/1.0)",
    "Accept": "application/json",
}

_QUOTE_CACHE_TTL_SECONDS = 4.0
_quote_cache: dict[str, tuple[float, float | None]] = {}
_quote_cache_lock = Lock()


def _yahoo_chart(symbol: str) -> dict | None:
    url = f"https://query1.finance.yahoo.com/v8/finance/chart/{symbol}"
    try:
        with httpx.Client(timeout=8.0, headers=YAHOO_HEADERS) as client:
            r = client.get(url)
            r.raise_for_status()
            return r.json()
    except Exception:
        return None


def symbol_exists_on_nse(name: str) -> bool:
    clean = name.strip().upper()
    if not clean or len(clean) > 50:
        return False
    data = _yahoo_chart(f"{clean}.NS")
    if not data:
        return False
    chart = data.get("chart") or {}
    result = chart.get("result")
    if not result:
        return False
    meta = (result[0] or {}).get("meta") or {}
    price = meta.get("regularMarketPrice") or meta.get("previousClose")
    return price is not None


def fetch_nifty_50_quote() -> tuple[float | None, float | None]:
    data = _yahoo_chart("^NSEI")
    if not data:
        return None, None
    chart = data.get("chart") or {}
    result = chart.get("result")
    if not result:
        return None, None
    meta = (result[0] or {}).get("meta") or {}
    price = meta.get("regularMarketPrice")
    prev = meta.get("previousClose") or meta.get("chartPreviousClose")
    if price is None or prev is None or prev == 0:
        return (float(price) if price is not None else None, None)
    change_pct = (float(price) - float(prev)) / float(prev) * 100.0
    return float(price), change_pct


def fetch_stock_quote_nse(name: str) -> float | None:
    """
    Returns current market price for an NSE equity symbol (NAME.NS).
    """
    clean = name.strip().upper()
    if not clean:
        return None
    data = _yahoo_chart(f"{clean}.NS")
    if not data:
        return None
    chart = data.get("chart") or {}
    result = chart.get("result")
    if not result:
        return None
    meta = (result[0] or {}).get("meta") or {}
    price = meta.get("regularMarketPrice") or meta.get("previousClose")
    return float(price) if price is not None else None


def get_quote_cached(name: str) -> float | None:
    """
    Short-TTL cache in front of fetch_stock_quote_nse so the price used to display
    (dashboard/execute reads) and the price used to auto-execute trades are always
    the same value, and so multiple callers within the TTL window don't each hit
    Yahoo's unofficial endpoint separately.
    """
    clean = name.strip().upper()
    now = time.monotonic()
    with _quote_cache_lock:
        cached = _quote_cache.get(clean)
        if cached is not None and now - cached[0] < _QUOTE_CACHE_TTL_SECONDS:
            return cached[1]
    price = fetch_stock_quote_nse(clean)
    with _quote_cache_lock:
        _quote_cache[clean] = (now, price)
    return price
